mm/context_label.py

```diff
@@ -0,0 +1,203 @@
+    ):
+        """
+        Get past/center/future context label of alignment by calculating back labels from dense tying and write the
+        labels into hdf file.
+        (C.f. NoStateTyingDense in rasr
+        https://github.com/rwth-i6/rasr/blob/a942e3940c30eeba900c873f3bfb3f48d5b39ddb/src/Am/ClassicStateTying.cc#L272)
+
+        :param alignment_cache_path: path to alginment cache
+        :param allophone_path: path to allohone
+        :param dense_tying_path: path to denser tying file
+        :param dense_label_info: the dense label information
+        :param sparse: writes the data to hdf in sparse format
+        :param returnn_root: path to returnn root
+        """
+        self.alignment_cache_path = alignment_cache_path
+        self.allophone_path = allophone_path
+        self.dense_tying_path = dense_tying_path
+        assert not (
+            dense_label_info.use_boundary_classes and dense_label_info.use_word_end_classes
+        ), "we do not use both class distinctions"
+        self.dense_label_info = dense_label_info
+        self.sparse = sparse
+        self.returnn_root = returnn_root
+
+        self.out_hdf_left_context = self.output_path("left_context.hdf")
+        self.out_hdf_right_context = self.output_path("right_context.hdf")
+        self.out_hdf_center_context = self.output_path("center_context.hdf")
+
+        self.rqmt = {"cpu": 1, "mem": 8, "time": 0.5}
+
+    def tasks(self):
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    @classmethod
+    def get_tying(cls, dense_tying_path: tk.Path) -> Tuple[Dict, int]:
+        with open(dense_tying_path.get_path()) as dense_tying_file:
+            state_tying = {
+                k: int(v) for l in dense_tying_file for k, v in [l.strip().split()[0:2]] if not l.startswith("#")
+            }
+
+        return state_tying
+
+    @classmethod
+    def get_target_labels_from_dense(cls, dense_label: int, dense_label_info: DenseLabelInfo) -> Tuple[int, int, int]:
+        num_boundary_classes = 4
+        num_word_end_classes = 2
+
+        future_label = np.mod(dense_label, dense_label_info.n_contexts)
+        pop_future_label = np.floor_divide(dense_label, dense_label_info.n_contexts)
+
+        past_label = np.mod(pop_future_label, dense_label_info.n_contexts)
+        center_state = np.floor_divide(pop_future_label, dense_label_info.n_contexts)
+
+        if dense_label_info.use_word_end_classes:
+            word_end_class = np.mod(center_state, num_word_end_classes)
+            center_state = np.floor_divide(center_state, num_word_end_classes)
+
+        if dense_label_info.use_boundary_classes:
+            boundary_class = np.mod(center_state, num_boundary_classes)
+            center_state = np.floor_divide(center_state, num_boundary_classes)
+
+        hmm_state_class = np.mod(center_state, dense_label_info.num_hmm_states_per_phon)
+        center_label = np.floor_divide(center_state, dense_label_info.num_hmm_states_per_phon)
```

Comment:
Why do you use `np.mod` and `np.floor_divice` and not just `%` and `//`? I think the latter is simpler and cleaner and better.

```suggestion
        future_label = dense_label % dense_label_info.n_contexts
        pop_future_label = dense_label // dense_label_info.n_contexts

        past_label = pop_future_label % dense_label_info.n_contexts
        center_state = pop_future_label // dense_label_info.n_contexts

        if dense_label_info.use_word_end_classes:
            word_end_class = center_state % num_word_end_classes
            center_state = center_state // num_word_end_classes

        if dense_label_info.use_boundary_classes:
            boundary_class = center_state % num_boundary_classes
            center_state = center_state // num_boundary_classes

        hmm_state_class = center_state % dense_label_info.num_hmm_states_per_phon
        center_label = center_state // dense_label_info.num_hmm_states_per_phon
```